### PR TITLE
Issue #32: Re-factoring of exception handling inside SIPpTest.post_run()

### PR DESCRIPTION
**Pull request by pcapdump**

### CODE REVIEW

File: sipplauncher/Test.py [hunk 1]
```diff
@@ -383,28 +382,41 @@ def post_run(self, run_id_prefix, args):
...
-                    self.__get_logger().debug(e, exc_info = True)
-                    state = SIPpTest.State.DIRTY
-
-            self.__set_state(state)
-            if state == SIPpTest.State.DIRTY:
+                    self.__run_script("after.sh", args)
+                finally:
+                    self.network.sniffer_stop()
+                    self.__remove_temp_folder(args)
+                    self.network.shutdown()
```
Comment (danielmartinezruiz) [inline sipplauncher/Test.py]: If `self.network.sniffer_stop()` raises an exception, `self.__remove_temp_folder(args)` and  `self.network.shutdown()` will not be executed. And I'm not sure, but I think the first exception will be lost and the catched exception outside (line 405) will be the last one.

Is this ok?

Comment (pcapdump) [reply sipplauncher/Test.py]: Great catch, Danny! I re-worked the implementation

File: sipplauncher/Run.py [hunk 2]
```diff
@@ -163,8 +163,15 @@ def run(args):
                 # 3. TestA run + TestB run
                 # 4. TestB post-run. OptionA: ValueB -> ValueA
                 # 5. TestA post-run: OptionB: ValueA -> ValueOrig
+                raise_exception = None
```
Comment (NGSegovia) [inline sipplauncher/Run.py]: This shouldn't be a list? In case of multiple tasks ending in an exception...

File: sipplauncher/Test.py [hunk 3]
```diff
@@ -398,13 +399,21 @@ def post_run(self, run_id_prefix, args):
                 except BaseException as e:
                     self.__get_logger().debug(e, exc_info = True)
                     state = SIPpTest.State.DIRTY
+                    if not isinstance(e, SIPpTest.ScriptRunException):
+                        # We should propagate exception to the caller if it's caused by internal error.
+                        # This stops tests execution.
+                        # We shouldn't propagae ScriptRunException, because it's caused by a test-suite content.
+                        # Therefore, it's not internal.
+                        raise_exception = e
```
Comment (danielmartinezruiz) [inline sipplauncher/Test.py]: If some errors occur on the loop not related to `ScriptRunException`, the last one will be raised to the upper catch block. Maybe it is more important to keep the first one, cause it should have the problem.

What about something like:

`if raise_exception is None and not isinstance(e, SIPpTest.ScriptRunException):`

?

File: sipplauncher/Run.py (hunk 2 above)
Comment (pcapdump) [reply sipplauncher/Run.py]: Nacho, in my understanding we can pass to the upper layer only 1 exception.
Currently we're passing the last exception.
So all other exceptions from the list will be unused.
Therefore, I don't see why should we store them, if they will be unused anyway.

File: sipplauncher/Test.py (hunk 3 above)
Comment (pcapdump) [reply sipplauncher/Test.py]: Danny, agree.
Will re-work it.

File: sipplauncher/Run.py (hunk 2 above)
Comment (NGSegovia) [reply sipplauncher/Run.py]: The problem I see is that we are returning just the last one, so we may missing important information. Can we at least log all exceptions? (if we are not logging them in other place).

Comment (pcapdump) [reply sipplauncher/Run.py]: Nacho, according to Danny's comment I changed the implementation to return just the first exception.

Regarding your comment, most of exceptions is already logged in test's run directory (try-except block inside SIPpTest.post_run()).
However some exceptions could be swallowed and hidden indeed (which don't hit the try-except block inside SIPpTest.post_run()).

The most simple solution would be to log the exception caught in Run.run() to main log as well.
This way, we'll store all exceptions in log, but for some cases there could be duplicates (same exception saved in test's run directory log + in main log).
To eliminate this logging dupication will need to have quite a big code change.
IMO it's not worth it.

So I propose just to add exception logging inside Run.run().

**Pull request merged**